ssl/ssl_sess.c

```diff
@@ -757,32 +757,36 @@ int SSL_CTX_add_session(SSL_CTX *ctx, SSL_SESSION *c)
         c->time = time(NULL);
         ssl_session_calculate_timeout(c);
     }
-    SSL_SESSION_list_add(ctx, c);
 
-    if (s != NULL) {
-        /*
-         * existing cache entry -- decrement previously incremented reference
-         * count because it already takes into account the cache
-         */
-
-        SSL_SESSION_free(s);    /* s == c */
-        ret = 0;
-    } else {
+    if (s == NULL) {
         /*
          * new cache entry -- remove old ones if cache has become too large
+	 * delete cache entry *before* add, so we don't remove the one we're adding!
```

Comment:
Weird indentation, possibly using `\t` ?

Reply:
I suppose this can be fixed when merging.

Reply:
I was using a different machine while writing this up, and my <TAB> settings were not what I thought they were.
I can fixup during merge.